### ISSUE TITLE
Intero can't find the project's local stack installation

### ISSUE THREAD

**Issue by dariooddenino** [hlissner/doom-emacs#434]
Hi! Honestly I'm not sure if this is a bug or something that just needs to be configured on the user side (in my noobness I tried for the last couple of hours without luck though).

1) Open doom-emacs.
2) Switch to a haskell project.
3) Open a haskell file.
4) I get an error saying intero can't find stack.

To make it work I have to launch emacs from the project root.
I'm on the latest commit on devel branch.

Thanks a lot :)

**Comment by hlissner** [comment]
Would you kindly run `DEBUG=1 make doctor`, copy its output into a pastebin, and post the link here? That will give me more insight into your issue.

**Comment by dariooddenino** [comment]
Here it is: https://pastebin.com/HCfXEsEd

**Comment by hlissner** [comment]
Hmm, nothing seems amiss on Doom's side so far. Could you try a couple things for me:

+ Run `M-x doom/info` then paste the results in here (it'll include some more information I need)
+ In Emacs, run `M-x eval-expression RET (executable-find "stack")` and let me know what that returns.
+ On your shell, run `which stack` and let me know what it outputs.

I suspect this is a shell issue. Would you happen to have your dotfiles up somewhere I can see them?

**Comment by dariooddenino** [comment]
1) The result of `M-x doom/info` is pasted at the bottom.
2) `M-x eval-expression RET (executable-find "stack")` returns nil. If I launch it from a folder that has a stack project it returns `~/.local/bin/stack`
3) `~/.local/bin/stack`

With spacemacs I have no problems on this same machine.
With dotfiles what do you mean specifically?

Thanks again :)

- OS: gnu/linux (x86_64-pc-linux-gnu)
- Emacs: 25.3.2 (set 12, 2017)
- Doom: 2.0.9 (develop https://github.com/hlissner/doom-emacs/commit/dc14738b0c53d4291c0fa1bbf412a7931f1ca704)
- Graphic display: t (daemon: nil)
- System features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GCONF GSETTINGS NOTIFY LIBSELINUX GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 MODULES
- Details:
  ```elisp
  modules:   (:config :feature eval evil file-templates (lookup (+devdocs +docsets)) services snippets spellcheck syntax-checker version-control workspaces :completion helm :ui doom-dashboard doom-modeline doom-quit hl-todo nav-flash evil-goggles vi-tilde-fringe window-select :tools electric-indent eshell gist imenu impatient-mode make neotree password-store pdf rotate-text term tmux upload :lang (haskell (+intero)) javascript markdown (org (+attach +babel +capture +export +present +publish)) purescript web :config)
  packages:  (ace-link ace-window all-the-icons async auth-password-store auto-compile auto-yasnippet avy (centered-window :recipe (centered-window :fetcher github :repo anler/centered-window-mode)) coffee-mode command-log-mode company company-dict company-ghc company-quickhelp company-statistics company-tern company-web devdocs dired-k doom-themes dumb-jump editorconfig (emacs-snippets :recipe (emacs-snippets :fetcher github :repo hlissner/emacs-snippets :files (*))) emmet-mode eslintd-fix evil evil-anzu evil-args evil-commentary evil-easymotion evil-embrace evil-escape evil-exchange evil-goggles evil-indent-plus evil-magit evil-matchit evil-mc evil-multiedit evil-numbers evil-org evil-snipe evil-surround evil-textobj-anyblock evil-vimish-fold evil-visualstar (exec-path-from-shell :ignore t) expand-region flycheck flycheck-cask flycheck-popup-tip flycheck-pos-tip flyspell-correct flyspell-correct-helm fringe-helper gist git-gutter-fringe git-link git-timemachine gitconfig-mode gitignore-mode haml-mode haskell-mode helm helm-ag helm-c-yasnippet helm-company helm-css-scss helm-dash (helm-describe-modes :recipe (helm-describe-modes :fetcher github :repo emacs-helm/helm-describe-modes)) helm-pass helm-projectile helm-swoop helm-xref helpful highlight-indentation highlight-numbers highlight-quoted hindent hl-todo htmlize hydra imenu-anywhere imenu-list impatient-mode intero js2-mode js2-refactor less-css-mode macrostep magit makefile-executor markdown-mode markdown-toc multi-term nav-flash neotree nlinum nlinum-hl nlinum-relative nodejs-repl ob-mongo ob-sql-mode ob-translate (org-bullets :recipe (org-bullets :fetcher github :repo Kaligule/org-bullets)) org-download org-plus-contrib org-tree-slide (osx-clipboard :ignore t) overseer ox-pandoc ox-reveal pass password-store pcre2el pdf-tools persistent-soft persp-mode prodigy projectile psc-ide pug-mode purescript-mode quelpa quickrun rainbow-delimiters rainbow-mode rjsx-mode (rotate-text :recipe (rotate-text :fetcher github :repo debug-ito/rotate-text.el)) sass-mode shackle shakespeare-mode shrink-path skewer-mode slime smart-forward smartparens solaire-mode ssh-deploy stylus-mode tern toc-org undo-tree use-package vi-tilde-fringe visual-fill-column web-beautify web-mode wgrep which-key xref-js2 yasnippet)
  elc dirs:  n/a
  exec-path: (/home/dario/.luna/bin /home/dario/.cargo/bin /home/dario/.local/share/umake/bin /home/dario/bin /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /usr/games /usr/local/games /snap/bin /usr/lib/x86_64-linux-gnu/emacs/25.3/x86_64-linux-gnu)
  ```

**Comment by hlissner** [comment]
Could you tell me what `$SHELL -c "which stack"` prints out on the command line?

> With dotfiles what do you mean specifically?

By dotfiles, I meant your bash/zsh config files, e.g. .bashrc, .bash_profile, .zshrc, .zshenv, etc.

**Comment by dariooddenino** [comment]
The result of `$SHELL -c "which stack" is `~/.local/bin/stack`. 
In spacemacs, if I run `(executable-find "stack")` I get that folder instead of `nil`.

My .bashrc
https://pastebin.com/N1M1y0jD

**Comment by hlissner** [comment]
I think I think I see the issue.

One thing left to try: could you open (Doom) Emacs from the terminal and see if `(executable-find "stack")` returns nil (when you aren't in a stack project)?

Also, how do you launch Emacs? Through an app launcher (like rofi, unity or desktop icon)? A terminal? Shell script?

**Comment by dariooddenino** [comment]
Yes, you're right: I just need to launch it from a terminal for the path to be there. 
I use rofi to launch emacs.

Adding that line didn't change anything :(

The confusing thing is that I launch spacemacs in the same exact way and it works fine. Maybe it's something they do in their config? I

**Comment by hlissner** [comment]
>Maybe it's something they do in their config?

You are exactly right. What spacemacs does differently is it uses `exec-path-from-shell` to proactively grab your PATH from an interactive shell session.

Doom only uses this on MacOS (but that's another story). On Linux, Doom does not, because it makes two assumptions:

1. You are running Emacs from the correct environment (which you are),
2. Your shell environment is configured properly for **non-interactive** shells (which yours is _not_)

In your bashrc, I see a lot of early `return`s if the session isn't interactive. In fact, only on line 5 you have:

```sh
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
```

This aborts your config before it gets to this important line on line 124:

```sh
PATH=$PATH:~/.local/bin
```

Thing is, Emacs is launched through in a non-interactive subshell, so Emacs inherits this gimped config.

I'd recommend putting `export PATH=~/.local/bin:$PATH` somewhere at the top, _before_ your early return for non-interactive sessions. As well as other PATH settings. That should fix your issue.

**Comment by dariooddenino** [comment]
I put it before that but it didn't work :( 
I will find a solution now that I know what the issue is.

Thanks a lot again!

**Comment by hlissner** [comment]
You may need a `.bash_profile` with this in it:

```sh
if [ -r ~/.bashrc ]; then
    . ~/.bashrc
fi
```

**Comment by dariooddenino** [comment]
It was already there. Tried half a dozen of different solutions I could find online without luck. In the end I created a custom startup script for doom that exports the path :)

**Comment by jeffjewiss** [comment]
Hi, I had a similar issue to this, but with MacOS and emacs not being able to find the executable for "tern". I could tell it was a PATH issue by similar troubleshooting outlined in this issue. If I open emacs from a terminal it works fine, but opening it using an application launcher (Alfred) causes PATH issues and error message repeatedly being displayed that tern cannot be found (since it's on a timer).

I was able to fix this by updating my paths in `/etc/paths`. I'm using "asdf" to manage Node versions and this is setup fine for zsh and bash, but only started working once I added the correct path to the above.

Just wanted to leave a note here in case anyone else runs into the same issue.

Cheers!
